### PR TITLE
Messaging 3.0

### LINKED ISSUE #3122: Upgrade to MP Reactive Operators 2.0
https://download.eclipse.org/microprofile/microprofile-reactive-streams-operators-2.0/microprofile-reactive-streams-operators-spec-2.0.html

- [x] Enable TCK tests
- [x] Update documentation and examples as needed

### LINKED ISSUE #3123: Upgrade to MP Reactive Messaging 3.0
### [MP Reactive Messaging 3.0](https://download.eclipse.org/microprofile/microprofile-reactive-messaging-3.0/microprofile-reactive-messaging-spec-3.0.html)

- [x] Emitter - asked for by almost every newcomer to MP Messaging(easier sending messages from imperative code)
- [x] Metrics
- [x] Negative acknowledgement(nacking) - redelivery based on nacks
- [x] Injection of channels to fields with @Channel 
- [x] Kafka connector @ConnectorAttribute - conf metadata for IDEs  #4430
- [x] JMS connector @ConnectorAttribute - conf metadata for IDEs #5327
- [x] AQ connector @ConnectorAttribute - conf metadata for IDEs #6035
- [x] Fix Kafka connector to use Nack #4430
- [x] Fix JMS connector to use Nack (defer) #4445
- [x] Enable TCK tests
- [x] Update documentation and examples as needed
- [x] Metrics #2105

### Blockers/related issues
- [x] eclipse/microprofile-reactive-messaging/issues/143
- [x] eclipse/microprofile-reactive-messaging/issues/145
- [x] eclipse/microprofile-reactive-messaging/issues/147
- [x] eclipse/microprofile-reactive-messaging/issues/148

### TCK pass progress
```
Tests run: 77, Failures: 0, Errors: 0, Skipped: 0
```

### PR DESCRIPTION
**Pull request by danielkec**
Fixes #3122 and partially #3123 
- all new messaging features implemented
- connectors not suppting nacking yet
- connectors still missing `@ConnectorAttribute`s 

**TCK tests for following issues are skipped:**
- [x] eclipse/microprofile-reactive-messaging/issues/143
- [x] eclipse/microprofile-reactive-messaging/issues/145
- [x] eclipse/microprofile-reactive-messaging/issues/147

### CODE REVIEW

File: microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/BufferedEmitter.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    public synchronized boolean isCancelled() {
+        return bep.isCancelled() || bep.isCompleted();
+    }
+
+    @Override
+    public boolean hasRequests() {
+        return bep.hasRequests();
+    }
+
+    public Publisher<?> getPublisher() {
```
Comment (tomas-langer) [inline microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/BufferedEmitter.java:75]: Either add `@Override` or remove public modifier

File: microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/BufferedEmitter.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        return FlowAdapters.toPublisher(bep);
+    }
+
+    public void validate(Object payload) {
```
Comment (tomas-langer) [inline microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/BufferedEmitter.java:79]: Either add `@Override` or remove public modifier

File: microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/LatestEmitter.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    public boolean hasRequests() {
+        return bep.hasRequests();
+    }
+
+    @Override
+    long getBufferLimit() {
+        return 3;
+    }
+
+    private synchronized Object dropFirst() {
```
Comment (tomas-langer) [inline microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/LatestEmitter.java:92]: 1. please do not use synchronized, replace with Lock
2. private methods should be last in class (public -> protected -> package private -> private)

File: microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/MessagingChannelProcessor.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        // noop
+    }
+
+    /**
+     * Messaging channel name associated with this processor.
+     * If the channel name is null this processor is applied to the all channel.
+     *
+     * @return Channel name or null
+     */
+    default String channelName() {
```
Comment (tomas-langer) [inline microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/MessagingChannelProcessor.java:112]: Development guidelines: never return null from public methods

File: microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/OutgoingMember.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.helidon.microprofile.messaging;
+
+import org.reactivestreams.Publisher;
+
```
Comment (tomas-langer) [inline microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/OutgoingMember.java:21]: A bit of a comment would be nice

File: microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/ProcessorMethod.java [hunk 6]
```diff
@@ -45,7 +56,7 @@ class ProcessorMethod extends AbstractMessagingMethod {
     }
 
     @Override
-    @SuppressWarnings("unchecked")
+    @SuppressWarnings({"unchecked", "checkstyle:methodlength"})
```
Comment (tomas-langer) [inline microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/ProcessorMethod.java:59]: do not suppress checkstyle warnings in code.
Use checkstyle-suppressions.xml instead if really needed

File: microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/ProcessorMethod.java [hunk 7]
```diff
@@ -54,34 +65,156 @@ public void init(BeanManager beanManager, Config config) {
...
+                                    return ReactiveStreams.fromCompletionStage(invoke(inMsg.getPayload()))
+                                            .onError(ackCtx::postNack)
+                                            .onErrorResumeWith(this::resumeOnError)
+                                            .peek(out -> ackCtx.postAck())
+                                            .map(MessageUtils::wrap);
+                                } catch (Throwable t) {
+                                    ackCtx.postNack(t);
+                                    return resumeOnError(t);
+                                }
+                            })
```
Comment (tomas-langer) [inline microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/ProcessorMethod.java:217]: missing a break after default

File: microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/ProcessorMethod.java [hunk 8]
```diff
@@ -54,34 +65,156 @@ public void init(BeanManager beanManager, Config config) {
...
+                                Message<?> inMsg = (Message<?>) in;
+                                AckCtx ackCtx = AckCtx.create(this, inMsg);
+                                try {
+                                    ackCtx.preAck();
+                                    return ReactiveStreams.fromPublisher(invoke(inMsg))
+                                            .onError(ackCtx::postNack)
+                                            .onErrorResumeWith(this::resumeOnError)
+                                            .peek(out -> ackCtx.postAck())
+                                            .map(MessageUtils::wrap);
+                                } catch (Throwable t) {
```
Comment (tomas-langer) [inline microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/ProcessorMethod.java:78]: the try/catch seems to be the same all the time. Could we extract at least that?

File: microprofile/messaging/core/src/test/java/io/helidon/microprofile/messaging/EmitterTest.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    @Inject
+    @Channel("kanal1")
+    private Emitter<String> emitter;
+
+    @Incoming("kanal1")
+    public void rec(String payload) {
+        System.out.println("Received " + payload);
+    }
+
+//    @Incoming("kanal1")
```
Comment (tomas-langer) [inline microprofile/messaging/core/src/test/java/io/helidon/microprofile/messaging/EmitterTest.java:57]: please clean up
test classes can be package private as well as test methods

File: microprofile/messaging/core/src/test/java/io/helidon/microprofile/messaging/inner/ChannelTest.java [hunk 10]
```diff
@@ -47,6 +47,7 @@ static Stream<CdiTestCase> testCaseSource() {
                         ChannelTest.class.getPackage().getName(),
                         ClassFilter.of(c -> Objects.nonNull(c.getAnnotation(ApplicationScoped.class))))
                 .stream()
+//                .filter(aClass -> aClass.getSimpleName().contains("ProcessorPublisherBuilderMsg2MsgPrepImplicitAckBean"))
```
Comment (tomas-langer) [inline microprofile/messaging/core/src/test/java/io/helidon/microprofile/messaging/inner/ChannelTest.java:50]: Clean up?

**Review by tomas-langer — CHANGES_REQUESTED** [review]

File: microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/ProcessorMethod.java (hunk 7 above)
Comment (danielkec) [reply microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/ProcessorMethod.java:217]: throws works the same

File: microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/MessagingChannelProcessor.java (hunk 4 above)
Comment (danielkec) [reply microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/MessagingChannelProcessor.java]: fixed

File: microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/OutgoingMember.java (hunk 5 above)
Comment (danielkec) [reply microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/OutgoingMember.java:21]: bit of comment added

File: microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/ProcessorMethod.java (hunk 6 above)
Comment (danielkec) [reply microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/ProcessorMethod.java]: supress removed

File: microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/ProcessorMethod.java (hunk 7 above)
Comment (danielkec) [reply microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/ProcessorMethod.java]: there is a throws in the default

File: microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/ProcessorMethod.java (hunk 8 above)
Comment (danielkec) [reply microprofile/messaging/core/src/main/java/io/helidon/microprofile/messaging/ProcessorMethod.java]: You are right, refactored to standalone method

File: microprofile/messaging/core/src/test/java/io/helidon/microprofile/messaging/inner/ChannelTest.java (hunk 10 above)
Comment (danielkec) [reply microprofile/messaging/core/src/test/java/io/helidon/microprofile/messaging/inner/ChannelTest.java]: fixed

**Review by tomas-langer — APPROVED** [review]
LGTM

**Pull request merged**